driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -221,9 +221,9 @@ public void onResult(final Void result, final Throwable t) {
                             LOGGER.trace(format("Pooled connection %s to server %s is now open",
                                                        pooledConnection.getDescription().getConnectionId(), serverId));
                         }
-                        callback.onResult(pooledConnection, null);
                         connectionPoolListener.connectionCheckedOut(
                                 new ConnectionCheckedOutEvent(pooledConnection.getDescription().getConnectionId()));
+                        callback.onResult(pooledConnection, null);
```

Comment:
Ordering of events as per the error case.

I initially thought the callback could have a side effect of closing the pooled connection, but it may just be a race between the whole pool being closed and the callback being called.